pennylane_runtime/vqe_runtime.py

```diff
@@ -0,0 +1,193 @@
+    ]
+
+    # Take the ansatz circuits and add measurements.
+    full_circs = [ansatz_circuit.compose(mcirc).measure_all(inplace=False) for mcirc in meas_circs]
+
+    num_params = ansatz_circuit.num_parameters
+
+    # Check initial state
+    if x0 is not None:
+        x0 = np.asarray(x0, dtype=float)
+        if x0.shape[0] != num_params:
+            shape = x0.shape[0]
+            raise ValueError(
+                f"Number of params in x0 ({shape}) does not match number \
+                              of ansatz parameters ({num_params})"
+            )
+    else:
+        x0 = 2 * np.pi * np.random.rand(num_params)
+
+    # Transpile the circuits
+    trans_dict = {}
+    if not backend.configuration().simulator:
+        trans_dict = {"layout_method": "sabre", "routing_method": "sabre"}
+    trans_circs = transpile(full_circs, backend, optimization_level=3, **trans_dict)
+
+    # Measurement mitigation
+    if use_measurement_mitigation:
+        maps = mthree.utils.final_measurement_mapping(trans_circs)
+        mit = mthree.M3Mitigation(backend)
+        mit.cals_from_system(maps)
+
+    def callback(*args):
+        user_messenger.publish(args)
+
+    def vqe_func(params):
+        # Binds parameters to the transpiled circuits.
+        bound_circs = [circ.bind_parameters(params) for circ in trans_circs]
+
+        # Submit the job and get the counts
+        counts = backend.run(bound_circs, shots=shots).result().get_counts()
+
+        if use_measurement_mitigation:
+            quasi_collection = mit.apply_correction(counts, maps)
+            expvals = quasi_collection.expval(meas_strings)
+        else:
+            expvals = mthree.utils.expval(counts, meas_strings)
+
+        energy = np.sum(coeffs * expvals).real
+        return energy
+
+    # SPSA and QNSPSA are taken from Qiskit and not SciPy
+    if optimizer == "SPSA":
+        spsa = SPSA(**optimizer_config, callback=callback)
+        x, loss, nfev = spsa.optimize(num_params, vqe_func, initial_point=x0)
+        res = OptimizeResult(
+            fun=loss,
+            x=x,
+            nit=optimizer_config["maxiter"],
+            nfev=nfev,
+            message="Optimization terminated successfully.",
+            success=True,
+        )
+    elif optimizer == "QNSPSA":
+        fidelity = QNSPSA.get_fidelity(ansatz_circuit)
+        spsa = QNSPSA(fidelity, **optimizer_config, callback=callback)
+        x, loss, nfev = spsa.optimize(num_params, vqe_func, initial_point=x0)
+        res = OptimizeResult(
+            fun=loss,
+            x=x,
+            nit=optimizer_config["maxiter"],
+            nfev=nfev,
+            message="Optimization terminated successfully.",
+            success=True,
+        )
+    # SciPy optimizers
+    else:
+        res = opt.minimize(
+            vqe_func, x0, method=optimizer, options=optimizer_config, callback=callback
+        )
+
+    return res
```

Comment:
I didn't know that you could design your own programs! I find it perfect 💯   Just out of curiosity, is there any kind of restriction on the libraries that can be used? you could have imported pennylane?

Reply:
You can only import libraries that are required by qiskit, so it is limiting. This is why I had to build functions to change the hamiltonian and circuits from PennyLane.